CHANGELOG.md

```diff
@@ -1,3 +1,8 @@
+## 0.12.15-dev
+
+* Add `package:matcher/expect.dart` library. Copies the implementation of
```

Comment:
I'm also not sure if the `matcher` target is used by any non `testonly` packages in google3

I suspect I'll need to keep `expect` as a separate target, so that the current deps on `matcher` don't pull in a transitive dep on `test_api`.

It's unfortunate that external pub solves will be impacted for non-dev dependencies on `matcher`, but that can't be avoided.

Reply:
Here is an example of a package that I think would be broken.

They export `matcher` and also export an `expect` method.

https://cs.github.com/eredo/cucumber_wire/blob/07a63c7d3bc933f84f3aede122174722791d4a8d/lib/cucumber_wire.dart

They do not currently constraint `matcher` to a single version, and so would be impacted.

https://cs.github.com/eredo/cucumber_wire/blob/07a63c7d3bc933f84f3aede122174722791d4a8d/pubspec.yaml#L18

Reply:
> To avoid impacting existing imports of `package:matcher/matcher.dart`. I wasn't sure if there were likely to be any import conflicts if anyone copied or duplicated the APIs outside of `test`.

If `test` is used in the same project, the version of matcher should also be constrained to not allow this version right? So because of that I think we would be safe.

I also think we could just bump matcher to `1.0` here if we wanted, which would further protect users. 

> I'm also not sure if the matcher target is used by any non testonly packages in google3

I would be surprised if this is the case, or if those couldn't be marked as test only also.

> I suspect I'll need to keep expect as a separate target, so that the current deps on matcher don't pull in a transitive dep on test_api.

Lets validate if this is the case? I would rather just keep the one `matcher.dart` import to eliminate confusion. It should make the migration easier if people can just import `package:matcher/matcher.dart`, which is likely to be the first thing they try.

> They do not currently constrain `matcher` to a single version, and so would be impacted.

Going to 1.0 would fix this right?

Reply:
> Going to 1.0 would fix this right?

It would prevent the error from surfacing, but also add churn. Existing uses would need to add `show` or `hide` to roll to the new version.

> If `test` is used in the same project, the version of matcher should also be constrained to not allow this version right? So because of that I think we would be safe.

Safe from the _original_ definition of these APIs, but not safe for _copies_ of the APIs like the various `expect` methods.

Reply:
> Lets validate if this is the case?

Its definitely the case today. I don't know how much work it would be to clean it up.

At a minimum `code_builder` is a blocker.
https://github.com/dart-lang/code_builder/blob/8cd7388121c55602df130bccc546f68476677e94/lib/src/matchers.dart#L5

I don't think we can make `code_builder` `testonly`, and it would be a breaking change to drop the dep on `matcher`.

Reply:
that is..... horrible

Reply:
Long term I really dislike the idea of having 2 imports for this package - especially when the one we want most people to use is _not_ the default one most people would import.

Reply:
I did file https://github.com/dart-lang/code_builder/issues/405 although I don't know if its worth a breaking change. But it really should be fixed.

Reply:
I don't _prefer_ the second unusual import. We're already in a strange place with imports and where APIs are defined, and wee might be missing an opportunity to make things better with this PR, but I don't necessarily think it is making the situation worse.

All of the impacted usage are, effectively, already legacy. I don't think it's worth spending time to clean up usages of a legacy package to maintain a single import, and then later clean up those usages to migrate them to `checks` or something else.

Reply:
In the ideal case, almost no one is going to ever write this import. Existing tests will get an analyzer aided import change, and new tests should prefer to use `checks`. I would hope it should be uncommon to add _new_ imports to `package:matcher/expect.dart` going forward.

Reply:
If we have an analyzer fix that will automatically add the correct import that alleviates some of my concern around this, but I do still think it leaves us in a bad place.

I foresee matcher existing for quite some time and continuing to get heavy use for quite some time? Or are we going to be more aggressively deprecating the package?

Reply:
Discussed separately - there is enough pain in trying to use a single library that it's not worth the effort.